SONiC Yang model support for NAT

AkhileshSamineni opened the pull request:
SONiC Yang model support for NAT

- What I did
Created SONiC Yang model for NAT
Tables: STATIC_NAPT, STATIC_NAT, NAT_GLOBAL, NAT_POOL, NAT_BINDINGS.

- How I did it
Defined Yang models for NAT based on Guideline doc:
https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md
and
https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md

- How to verify it
Added test cases to verify it.

Signed-off-by: Akhilesh Samineni <akhilesh.samineni@broadcom.com>

venkatmahalingam commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,281 @@
+        "SONiC NAT";
+
+    revision 2019-10-30 {
+        description
```
Update the revision to reflect the latest date.

AkhileshSamineni replied:
Done.

venkatmahalingam commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,281 @@
+        input {
+            leaf nat-param {
+                type nat-clear-param;
+                description
```
Not sure we have any use-case with rpc in the community today, if not required, can this be removed?

AkhileshSamineni replied:
Will remove the rpc section.

venkatmahalingam commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,281 @@
+            list STATIC_NAPT_LIST {
+                key "global_ip ip_protocol global_l4_port";
+
+                ext:key-regex-configdb-to-yang "^([a-fA-F0-9:./]+)|(TCP|UDP)|([0-9]+)$";
```
I believe, we decided not to define key-regex for the list, @praveen-li We don't need this anymore..correct?

AkhileshSamineni replied:
Removed key-regex.

AkhileshSamineni commented:
/Azurepipelines run

lguohan commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+                        pattern '[a-zA-Z0-9]{1}([-a-zA-Z0-9_]{0,31})';
+                        length 1..32 {
+                            error-app-tag pool-name-invalid-length;
+                        }
```
all fields needs descriptions

AkhileshSamineni replied:
Added descriptions for all the fields.

lguohan commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+            list NAT_POOL_LIST {
+                key "pool_name";
+
+                leaf pool_name {
```
change pool_name to name

AkhileshSamineni replied:
Done.

lguohan commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+
+                leaf nat_port {
+                    type string {
+                        pattern "[0-9]{1,5}(-)[0-9]{1,5}";
```
must be valid range

AkhileshSamineni replied:
Added valid range.

lguohan commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+            }
+        }
+
+        container NAT_POOL {
```
max elements?

AkhileshSamineni replied:
Added max-elements.

lguohan commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+        }
+    }
+
+    typedef ip-protocol-type {
```
move sonic-types.yang

AkhileshSamineni replied:
Done.

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+        }
+    }
+
+    typedef ip-address {
```
Change it to ip-address range.

AkhileshSamineni replied:
Done.

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+
+    container sonic-nat {
+
+        container STATIC_NAPT {
```
Add descriptions.

AkhileshSamineni replied:
Done.

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+                }
+
+                leaf nat_ip {
+                    type ip-address;
```
Add mandatory here.

AkhileshSamineni replied:
Done.

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+                }
+
+                leaf nat_port {
+                    type string {
```
Use proper port numbers.

AkhileshSamineni replied:
Done.

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+            list NAT_POOL_LIST {
+                key "pool_name";
+
+                leaf pool_name {
```
Change the pool_name to name.

AkhileshSamineni replied:
Done,

AkhileshSamineni commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+
+                        pattern '[a-zA-Z0-9]{1}([-a-zA-Z0-9_]{0,31})';
+                        length 1..32 {
+                            error-app-tag binding-name-invalid-length;
```
Add error-message.

AkhileshSamineni replied:
Done.

arlakshm commented on src/sonic-yang-models/tests/files/sample_config_db.json:
```diff
@@ -682,6 +682,39 @@
                 "nexthop_group_threshold_type": "percentage",
                 "polling_interval": "0"
             }
+        },
```
Nat zone is not needed in yang model?

AkhileshSamineni replied:
Yes, added NAT zone for INTERFACE, VLAN_INTERFACE and LOOPBACK_INTERFACE tables.

praveen-li commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,284 @@
+
+        container STATIC_NAPT {
+
+            list STATIC_NAPT_LIST {
```
Max element in a yang list was part of RFC 6020, so I think we do not need to specify any version for this. But as I know there was a bug in Libyang library related to Max elements. 
Make sure that, you are not hitting the same bug.

AkhileshSamineni replied:
Yes, added max-elements to NAT_POOL, NAT_BINDINGS and NAT_GLOBAL tables.

AkhileshSamineni commented:
@praveen-li @arlakshm @lguohan @venkatmahalingam
Please re-review the changes.
I have addressed all the review comments and added test cases as well.

venkatmahalingam commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+
+            description "NAT_GLOBAL table part of config_db.json";
+
+            list NAT_GLOBAL_LIST {
```
Replace list (with max-elements 1) with a container? e.g container NAT_GLOBAL { container values {...}}

AkhileshSamineni replied:
Seems like List is mandatory here, seeing errors when I replace list with container.
Is it supported ? If yes, could you please give me some reference here.

venkatmahalingam replied:
Please refer  TACPLUS container in file  src/sonic-yang-models/yang-models/sonic-system-tacacs.yang 
PR - https://github.com/Azure/sonic-buildimage/pull/7671/files

AkhileshSamineni replied:
Done, replaced list with a container.

venkatmahalingam commented:
@AkhileshSamineni Please fix the build failures.

AkhileshSamineni commented:
> @AkhileshSamineni Please fix the build failures.

@venkatmahalingam
Build failure is due to Port breakout failure in "sonic_utilities-1.2-py3-none-any.whl"
```
[ERROR] Port breakout Failed!!! Opting Out
Failed to break out Port. Error: 
```
Do you have any idea, why we are seeing this failure ?, please let me know.

AkhileshSamineni commented:
@praveen-li  @arlakshm  @lguohan @venkatmahalingam 
This time Build is passed in Broadcom platform, it is failing on Mellanox and VS.
Could someone check it and help me out.

arlakshm commented:
@kperumalbfn, can you please take a look at this PR

arlakshm commented:
@AkhileshSamineni the sonic-utilities build is failing with the below error.
Can you check.
```

Failed to validate data tree
Data Validation Failed
[ERROR] Port breakout Failed!!! Opting Out
Failed to break out Port. Error: 

```

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+                    description "Admin mode of the NAT feature.";
+
+                    type stypes:admin_mode;
+                    default enabled;
```
Should the default mode be disabled?

AkhileshSamineni replied:
Yes, corrected.

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+
+                leaf nat_timeout {
+                    description "Timeout for the nat entries within the range of 300 sec to 432000 secs.";
+
```
Can we specify default values for both TCP and UDP timeout?

AkhileshSamineni replied:
Done, specified default values for timeout.

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+                    type leafref {
+                        path "../../../NAT_POOL/NAT_POOL_LIST/name";
+                    }
+                }
```
We need to update DO_NOT_NAT_ACTION in ACL Yang model?

AkhileshSamineni replied:
Done, added DO_NOT_NAT as action.

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+                key "global_ip";
+
+                leaf global_ip {
+                    description "Global ip for the Static NAT entry.";
```
mandatory global_ip

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-nat.yang:
```diff
@@ -0,0 +1,316 @@
+
+                leaf global_ip {
+                    description "Global ip for the Static NAPT entry.";
+
```
mandatory global_ip, protocol and l4_port

AkhileshSamineni replied:
Not needed to mention mandatory keyword for global_ip, protocol and l4_port, as by default key parameters are mandatory.

kperumalbfn commented on src/sonic-yang-models/yang-models/sonic-vlan.yang:
```diff
@@ -61,6 +61,16 @@ module sonic-vlan {
+
+				leaf nat_zone {
+					description "NAT Zone for the vlan interface";
+					type uint8 {
```
default value of nat-zone 0?

AkhileshSamineni replied:
Added default values.

venkatmahalingam commented:
@AkhileshSamineni Please mark the comments addressed as resolved to track the outstanding comments better,

AkhileshSamineni commented:
> @AkhileshSamineni Please mark the comments addressed as resolved to track the outstanding comments better,

@venkatmahalingam I have addressed all the comments and mentioned those as resolved.

AkhileshSamineni commented:
@praveen-li   @arlakshm  @lguohan  @venkatmahalingam 
The build failure - [ERROR] Port breakout Failed!!! Opting Out
Not sure why I am seeing this issue, Someone give me some pointers here please.
sonic-utilities build is failing sometimes and passing sometimes.

praveen-li commented:
I will have a look today, Will it be possible for you to send the latest diff of your PR to me.



Thanks,

Regards
Praveen
________________________________
From: Akhilesh Samineni ***@***.***>
Sent: Monday, May 3, 2021 4:03 AM
To: Azure/sonic-buildimage ***@***.***>
Cc: Praveen Chaudhary ***@***.***>; Mention ***@***.***>
Subject: Re: [Azure/sonic-buildimage] SONiC Yang model support for NAT (#7051)


@praveen-li<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpraveen-li&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219078353%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nmV%2Fs7qlsHR0uEgBsh%2FbIxBUp%2F9PSBQES%2Btj5FK1L3w%3D&reserved=0> @arlakshm<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Farlakshm&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219088313%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=5jtvP%2Ff0W5JHIkDO2Bk6%2FoAwh8BfjYsKTM%2Bd9MLjtms%3D&reserved=0> @lguohan<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flguohan&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219088313%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PwHQXPI5tIToMQPKl%2FZTnibg3oi7Na0tN3%2Fe7Xh8ynk%3D&reserved=0> @venkatmahalingam<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fvenkatmahalingam&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=WQbLJZwilKNmjAQtqIu5Rt0jytAzC26%2Fkfv4GTzPB%2F4%3D&reserved=0>
The build failure - [ERROR] Port breakout Failed!!! Opting Out
Not sure why I am seeing this issue, Someone give me some pointers here please.
sonic-utilities build is failing sometimes and passing sometimes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsonic-buildimage%2Fpull%2F7051%23issuecomment-831188438&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=3YOfnoBckveOHC57dPkh4x2x9XnPUPp9jASCsfawC24%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJHZHPPYADBRCHOAMLPOZCLTLZ7IRANCNFSM4ZDWYQFA&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=uj350JOgDYnMOi0mOsJkyiUzLb37PMl8PjKTUqnlmWY%3D&reserved=0>.

AkhileshSamineni commented:
Attached the patch file for sonic-yang changes for NAT.

Thanks,
Akhilesh

On Mon, May 3, 2021 at 9:36 PM Praveen Chaudhary ***@***.***>
wrote:

> I will have a look today, Will it be possible for you to send the latest
> diff of your PR to me.
>
>
>
> Thanks,
>
> Regards
> Praveen
> ________________________________
> From: Akhilesh Samineni ***@***.***>
> Sent: Monday, May 3, 2021 4:03 AM
> To: Azure/sonic-buildimage ***@***.***>
> Cc: Praveen Chaudhary ***@***.***>; Mention ***@***.***>
> Subject: Re: [Azure/sonic-buildimage] SONiC Yang model support for NAT
> (#7051)
>
>
> @praveen-li<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fpraveen-li&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219078353%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nmV%2Fs7qlsHR0uEgBsh%2FbIxBUp%2F9PSBQES%2Btj5FK1L3w%3D&reserved=0>
> @arlakshm<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Farlakshm&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219088313%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=5jtvP%2Ff0W5JHIkDO2Bk6%2FoAwh8BfjYsKTM%2Bd9MLjtms%3D&reserved=0>
> @lguohan<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flguohan&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219088313%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PwHQXPI5tIToMQPKl%2FZTnibg3oi7Na0tN3%2Fe7Xh8ynk%3D&reserved=0>
> @venkatmahalingam<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fvenkatmahalingam&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=WQbLJZwilKNmjAQtqIu5Rt0jytAzC26%2Fkfv4GTzPB%2F4%3D&reserved=0
> >
> The build failure - [ERROR] Port breakout Failed!!! Opting Out
> Not sure why I am seeing this issue, Someone give me some pointers here
> please.
> sonic-utilities build is failing sometimes and passing sometimes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsonic-buildimage%2Fpull%2F7051%23issuecomment-831188438&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=3YOfnoBckveOHC57dPkh4x2x9XnPUPp9jASCsfawC24%3D&reserved=0>,
> or unsubscribe<
> https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJHZHPPYADBRCHOAMLPOZCLTLZ7IRANCNFSM4ZDWYQFA&data=04%7C01%7Cpchaudhary%40linkedin.com%7C4776cae3499e4dcff44208d90e231be6%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637556366219098266%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=uj350JOgDYnMOi0mOsJkyiUzLb37PMl8PjKTUqnlmWY%3D&reserved=0
> >.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Azure/sonic-buildimage/pull/7051#issuecomment-831363618>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALLTGRGRDWGRZD2MN4RRE7DTL3CWZANCNFSM4ZDWYQFA>
> .
>

-- 
This electronic communication and the information and any files transmitted 
with it, or attached to it, are confidential and are intended solely for 
the use of the individual or entity to whom it is addressed and may contain 
information that is confidential, legally privileged, protected by privacy 
laws, or otherwise restricted from disclosure to anyone else. If you are 
not the intended recipient or the person responsible for delivering the 
e-mail to the intended recipient, you are hereby notified that any use, 
copying, distributing, dissemination, forwarding, printing, or copying of 
this e-mail is strictly prohibited. If you received this e-mail in error, 
please return the e-mail to the sender, delete it from your computer, and 
destroy any printed copy of it.

AkhileshSamineni commented:
@praveen-li  @arlakshm  @lguohan  @venkatmahalingam 
I have addressed all the review comments, Please merge this PR.

AkhileshSamineni commented:
@arlakshm @lguohan Can someone merge this PR please.

arlakshm commented:
@AkhileshSamineni can you address the latest comment from @venkatmahalingam ?

AkhileshSamineni commented:
> @AkhileshSamineni can you address the latest comment from @venkatmahalingam ?

@arlakshm Addressed the review comment.
